elastic_agent_client/service/checkin.py

```diff
@@ -42,27 +54,51 @@ async def _run(self):
         send_checkins_task.add_done_callback(functools.partial(self._callback))
         receive_checkins_task.add_done_callback(functools.partial(self._callback))
 
-        try:
-            logger.debug(f"Running {self.name} service loop")
-            await asyncio.wait([send_checkins_task, receive_checkins_task])
-        except Exception:
-            send_checkins_task.cancel()
-            receive_checkins_task.cancel()
-            raise
+        self._send_checkins_task = send_checkins_task
+        self._receive_checkins_task = receive_checkins_task
+
+        logger.debug(f"Running {self.name} service loop")
+        done, pending = await asyncio.wait(
+            [send_checkins_task, receive_checkins_task],
+            return_when=asyncio.FIRST_EXCEPTION,
+        )
+
+        for task in pending:
+            task.cancel()
+            try:
+                await task
+            except CancelledError:
+                logger.error("Task did not handle cancellation gracefully")
+
+        # Separated these two to log all errors if both tasks error out together
+        # Which is unlikely, but it's cheap to do it the way I did
+        for task in done:
+            if not task.cancelled() and task.exception():
+                logger.error(
+                    f"Task {task.get_name()} terminated due to exception:",
+                    exc_info=task.exception(),
+                )
+
+        for task in done:
+            if not task.cancelled():
+                task_exception = task.exception()
+                if task_exception:
+                    raise task_exception
 
     async def send_checkins(self, send_queue):
         while self.running:
             if send_queue.empty():
                 await self.do_checkin(send_queue)
-            await sleep(self.CHECKIN_INTERVAL)
+            # To not sleep if not running any more
```

Comment:
Super nit: this comment actually confused me with the double negative. I think it could be clearer

```suggestion
            # Sleep if still running
```